core/integration/storagetest/mutation_log.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storagetest
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/google/keytransparency/core/keyserver"
+
+	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+)
+
+type QueueStorageFactory func(ctx context.Context, t *testing.T, dirID string, logIDs ...int64) keyserver.MutationLogs
+
+type QueueStorageTest func(ctx context.Context, t *testing.T, f QueueStorageFactory)
+
+// RunQueueStorageTests runs all the queue tests against the provided storage implementation.
+func RunQueueStorageTests(t *testing.T, factory QueueStorageFactory) {
+	ctx := context.Background()
+	b := &QueueTests{}
+	for name, f := range map[string]QueueStorageTest{
+		// TODO(gbelvin): Discover test methods via reflection.
+		"TestReadLog": b.TestReadLog,
+	} {
+		t.Run(name, func(t *testing.T) { f(ctx, t, factory) })
+	}
+}
+
+// QueueTests is a suite of tests to run against
+type QueueTests struct{}
+
+func mustMarshal(t *testing.T, p proto.Message) []byte {
+	t.Helper()
+	b, err := proto.Marshal(p)
+	if err != nil {
+		t.Fatalf("proto.Marshal(): %v", err)
+	}
+	return b
+}
+
+func (QueueTests) TestReadLog(ctx context.Context, t *testing.T, newForTest QueueStorageFactory) {
```

Comment:
Given that these tests are performing high level conformance of some fairly extensive iceberg code, I suggest adding some documentation on what this is testing and why is reasonable. Justification: $randomEngineer doesn't know a lot about this & updates their DB from v1.2 to v1.3; the batchSize 4 test below fails and prints "ReadLog(4): len 12, want 6". Poor engineer then has no idea what the devil is going on and goes to the pub. Tests stay broken.